### PR TITLE
Cut test suite runtime from 27 min to 14 min on CircleCI

### PR DESCRIPTION
**Pull request by AlexNisnevich**
Thanks to @money for suggesting some of these techniques.

- Enable profiling in rspec to track which specs are taking the longest.
- Turn `before :each` blocks into `let` declarations whenever possible.
- `build` factories instead of `create`ing them whenever possible.
- Create incidents using factories instead of requests in all cases where we don't care about the incident creation itself. (<- this is what led to most of the speedup)
- Add rubocop rules to avoid bad behavior (in particular, using instance variables in specs) in the future.

I think it's possible to get the tests even faster with some more of the same work, but this is a decent starting point.

### CODE REVIEW

File: spec/factories/incident.rb [hunk 1]
```diff
@@ -17,15 +20,19 @@
         )
       end
 
-      e.num_civilians.times do
-        incident.involved_civilians << create(:involved_civilian)
-      end
+      unless e.stop_step == :civilians
```
Comment (wetchler) [inline spec/factories/incident.rb:23]: ooo this stop_step constructor is very clever and useful and i like it

**Comment by money** [comment]
This is great @AlexNisnevich! More than 50% improvement with changing 3 things.

File: app/models/concerns/serializable.rb [hunk 2]
```diff
@@ -46,7 +46,9 @@ def xml_schema
 
       def displayed_type(field, type)
         # Get a string of options for this field, if Incident::OPTIONS_PER_FIELD[field] is defined.
-        options = Incident::OPTIONS_PER_FIELD[field].try { |opts| opts.map { |o| "'#{o}'" }.join(', ') }
+        options = Incident::OPTIONS_PER_FIELD[field].try do |opts|
+          opts.map { |o| "'#{o}'" }.join(', ').sub("&", "&amp;")
```
Comment (wetchler) [inline app/models/concerns/serializable.rb:50]: So I just had a thought -- the issue with this is that it's rendered as HTML right? In that case, maybe we should send it through a general html-ify thing? In case other characters (e.g. `<`) appear in the future

File: spec/mailers/bridge_mailer_spec.rb [hunk 3]
```diff
@@ -14,33 +14,31 @@ def extract_address_only(mail_target)
   end
 
   describe 'feedback_email' do
-    before :all do
-      @user = create :dummy_user
-      @feedback = Feedback.create(source: "Source test", content: "Content test")
-      @mail = BridgeMailer.feedback_email(@feedback, @user).deliver_now
-    end
+    let(:user) { build :dummy_user }
+    let(:feedback) { Feedback.create(source: "Source test", content: "Content test") }
```
Comment (wetchler) [inline spec/mailers/bridge_mailer_spec.rb]: Does this need to be `create` or would `build` do?

File: spec/models/agency_status_spec.rb [hunk 4]
```diff
@@ -1,45 +1,44 @@
...
-    @ori = "ORI_1"
-    @status = AgencyStatus.create!(ori: @ori)
-    @other_ori = "ORI_2"
-    @other_status = AgencyStatus.create!(ori: @other_ori)
-  end
+describe AgencyStatus, type: :model do
+  let(:ori) { "ORI_1" }
+  let(:other_ori) { "ORI_2" }
+
+  let!(:status) { AgencyStatus.create!(ori: ori) }
```
Comment (wetchler) [inline spec/models/agency_status_spec.rb]: Any reason for `create!` instead of `create` here and below? It seems to me we should use one or the other throughout the specs, no?

File: spec/models/general_info_spec.rb [hunk 5]
```diff
@@ -1,42 +1,40 @@
 require 'rails_helper'
 
-describe '[General Info]', type: :model do
+describe GeneralInfo, type: :model do
   describe '[Validation]' do
-    before :each do
-      @info = create :general_info
-    end
+    let(:gi) { build :general_info }
```
Comment (wetchler) [inline spec/models/general_info_spec.rb:5]: Wait.... I thought `let` prevented you from changing the variable? But below you change `gi` all the time? Or does it just return the same reference to the same object, so you can't redefine the variable (e.g. `gi = 5`)?

File: spec/requests/devise_signup_spec.rb [hunk 6]
```diff
@@ -32,28 +32,26 @@ def visit_registration_and_fill_form
   end
 
   describe '[with an existing user]' do
-    before :each do
-      @dummy_user = create :dummy_user
-    end
+    let(:dummy_user) { create :dummy_user }
```
Comment (wetchler) [inline spec/requests/devise_signup_spec.rb]: This should really be `let!` if it's something we want created before anything runs. In our case it will work anyway because of the way it happens to be called, but `let!` is designed for this purpose

File: spec/requests/doj_dashboard_spec.rb [hunk 7]
```diff
@@ -1,25 +1,24 @@
...
+      build(:dummy_user, first_name: "Elvis", last_name: "Presley",
+                         email: "elvis@example.com", role: Rails.configuration.x.roles.doj,
+                         user_id: 'elvis', ori: 'ALPHA_ORI')
     end
 
-    describe '[logged in as a doj user]' do
+    describe '[logged in as a DOJ user]' do
       before :each do
-        login user: @elvis
+        login user: elvis
```
Comment (wetchler) [inline spec/requests/doj_dashboard_spec.rb:18]: Given that we only `build` the user, I was surprised this works. I think this only works because we are using the Siteminder login system, where "logging in" will create a user if there is none. If we were using Devise, login would fail because the user doesn't exist.

If I'm thinking correctly, we should use `let!(:elvis)` with `create`

File: spec/requests/doj_dashboard_spec.rb [hunk 8]
```diff
@@ -66,17 +65,21 @@
         GlobalState.open_submission_window!
```
Comment (wetchler) [inline spec/requests/doj_dashboard_spec.rb:65]: Also, this can be a `before :all` block, no?

File: spec/requests/incident_id_spec.rb [hunk 9]
```diff
@@ -1,14 +1,17 @@
...
 describe '[Incident ID]', type: :request do
+  let(:user) { build :dummy_user, ori: 'CA0123456' }
+
   before :all do
     Rails.configuration.x.branding.incident_id_prefix = 'PREFIX'
   end
 
-  it 'generates valid ursus id after general_info step' do
-    user = build :dummy_user, ori: 'CA0123456'
+  before :each do
```
Comment (wetchler) [inline spec/requests/incident_id_spec.rb:10]: Why not `before :all`?

File: spec/requests/incident_spec.rb [hunk 10]
```diff
@@ -280,108 +274,106 @@ def breadcrumbs_classes
...
+      expect(page).to have_no_content('Female')
 
       find('h1', text: 'Officer').find('~a', text: 'Edit').click
       expect(current_path).to end_with('/involved_officers/0/edit')
-      check 'Cambodian'
+      choose 'Female'
       click_button 'Save and Continue'
-      expect(page).to have_content('White')
-      expect(page).to have_content('Cambodian')
+      expect(page).to have_content('Female')
```
Comment (wetchler) [inline spec/requests/incident_spec.rb:359]: Why these changes?

File: spec/requests/involved_civilian_spec.rb [hunk 11]
```diff
@@ -209,27 +211,21 @@
...
-    # Uncomment for debugging
-    # (0..9).each do |i|
-    #   puts "Unsorted #{i} #{incident.involved_civilians[i].id} #{incident.involved_civilians[i].created_at}"
-    #   puts "Sorted   #{i} #{civs_sorted[i].id} #{civs_sorted[i].created_at}"
-    # end
-    (1..9).each do |i|
-      expect(civs_sorted[i].created_at).to be > civs_sorted[i - 1].created_at
-    end
+    civilians_sorted = InvolvedPerson.sorted_persons(incident.involved_civilians)
+    expect(civilians_sorted.map(&:created_at)).to be_sorted
```
Comment (wetchler) [inline spec/requests/involved_civilian_spec.rb:229]: Whoa.... TIL be_sorted

File: spec/requests/doj_dashboard_spec.rb [hunk 12]
```diff
@@ -1,25 +1,24 @@
...
   end
 
   describe '[with some mock data]' do
-    before :each do
-      stub_const('Constants::DEPARTMENT_BY_ORI', 'ALPHA_ORI' => 'ALPHA_DEPT', 'BRAVO_ORI' => 'BRAVO_DEPT')
-      @elvis = create(:dummy_user, first_name: "Elvis", last_name: "Presley",
-                                   email: "elvis@example.com", role: Rails.configuration.x.roles.doj,
-                                   user_id: 'elvis', ori: 'ALPHA_ORI')
+    let(:elvis) do
+      build(:dummy_user, first_name: "Elvis", last_name: "Presley",
```
Comment (wetchler) [inline spec/requests/doj_dashboard_spec.rb]: Since users have to be created before you can log in as them, there isn't much advantage to using `build` here. I'd actually argue that it would be clearer to use a `let!` block that calls `create`

File: spec/models/agency_status_spec.rb (hunk 4 above)
Comment (AlexNisnevich) [reply spec/models/agency_status_spec.rb]: No reason, I'll just make it `create` for consistency.

File: spec/mailers/bridge_mailer_spec.rb (hunk 3 above)
Comment (AlexNisnevich) [reply spec/mailers/bridge_mailer_spec.rb]: Let me check and see, `build` ought to be fine.

File: spec/factories/incident.rb (hunk 1 above)
Comment (AlexNisnevich) [reply spec/factories/incident.rb:23]: 😄

File: app/models/concerns/serializable.rb (hunk 2 above)
Comment (AlexNisnevich) [reply app/models/concerns/serializable.rb:50]: Hm, so I'm actually going to take this out of this PR because it was addressed in a previous PR. I think you're right and maybe we should make a low-priority issue to do that.

File: spec/models/general_info_spec.rb (hunk 5 above)
Comment (AlexNisnevich) [reply spec/models/general_info_spec.rb:5]: It just returns the same reference to the same object, so I can't redefine `gi` but I can work with it the way I do.

File: spec/requests/doj_dashboard_spec.rb (hunk 12 above)
Comment (AlexNisnevich) [reply spec/requests/doj_dashboard_spec.rb]: Agreed, will change.

File: spec/requests/devise_signup_spec.rb (hunk 6 above)
Comment (AlexNisnevich) [reply spec/requests/devise_signup_spec.rb]: Agreed, will change.

File: spec/requests/doj_dashboard_spec.rb (hunk 7 above)
Comment (AlexNisnevich) [reply spec/requests/doj_dashboard_spec.rb:18]: Agreed, will change.

File: spec/requests/doj_dashboard_spec.rb (hunk 8 above)
Comment (AlexNisnevich) [reply spec/requests/doj_dashboard_spec.rb:65]: I think so, let me test that.

File: spec/requests/incident_id_spec.rb (hunk 9 above)
Comment (AlexNisnevich) [reply spec/requests/incident_id_spec.rb:10]: I don't think `login` would work inside a `before :all`, but I can try it.

File: spec/requests/incident_spec.rb (hunk 10 above)
Comment (AlexNisnevich) [reply spec/requests/incident_spec.rb:359]: Race is nondeterministic when we use the `incident` factory, but gender is one of the deterministic fields.

File: spec/requests/involved_civilian_spec.rb (hunk 11 above)
Comment (AlexNisnevich) [reply spec/requests/involved_civilian_spec.rb:229]: Actually I just implemented `Array.sorted?` in `lib/core_extensions/array.rb`, but I do think this ought to be a built-in feature of Ruby / ActiveSupport.

File: spec/requests/doj_dashboard_spec.rb (hunk 8 above)
Comment (AlexNisnevich) [reply spec/requests/doj_dashboard_spec.rb:65]: No, it looks like that won't work:

```
 RuntimeError:
       let declaration `elvis` accessed in a `before(:context)` hook at:
         /bridge/spec/requests/doj_dashboard_spec.rb:76:in `block (4 levels) in <top (required)>'

       `let` and `subject` declarations are not intended to be called
       in a `before(:context)` hook, as they exist to define state that
       is reset between each example, while `before(:context)` exists to
       define state that is shared across examples in an example group
```

RSpec is picky about what can go inside a `before :all` hook, which makes sense given how much state is reset between tests.

File: spec/requests/incident_id_spec.rb (hunk 9 above)
Comment (AlexNisnevich) [reply spec/requests/incident_id_spec.rb:10]: Yeah, see my reply to your comment on doj_dashboard_spec.rb.

**Comment by AlexNisnevich** [comment]
Ready for re-review, @everett-wetchler .

**Review by wetchler — APPROVED** [review]
